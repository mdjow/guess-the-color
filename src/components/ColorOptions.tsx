import styled from "styled-components";

interface ColorOptionsProps {
  options: string[];
  disabled: boolean;
  onSelect: (color: string) => void;
}

export const ColorOptions = ({
  onSelect,
  options,
  disabled,
}: ColorOptionsProps) => (
  <ColorOptionsWrapper>
    {options.map((color, i) => (
      <Option
        key={i}
        onClick={() => onSelect(color)}
        color={color}
        disabled={disabled}
      >
        {color}
      </Option>
    ))}
  </ColorOptionsWrapper>
);

const ColorOptionsWrapper = styled.div`
  display: flex;
  gap: 0.3rem;
`;

const Option = styled.button`
  width: 100%;
  padding: 1rem;
  font-size: 1rem;
  font-weight: 600;
  border-radius: var(--border-radius);
  cursor: pointer;
  background-color: #ffffff;
  border: 1px solid #e0e0e0;

  &:hover:enabled {
    background-color: #eeeeee;
  }
`;
