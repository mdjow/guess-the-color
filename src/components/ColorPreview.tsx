import styled from "styled-components";

interface ColorPreviewProps {
  color: string;
  started: boolean;
  onStart: () => void;
}

export const ColorPreview = ({
  color,
  started,
  onStart,
}: ColorPreviewProps) => (
  <ColorPreviewWrapper color={color}>
    {!started && <StartButton onClick={onStart}>Start</StartButton>}
  </ColorPreviewWrapper>
);

const ColorPreviewWrapper = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 20rem;
  background-color: ${(props) => props.color};
`;

const StartButton = styled.button`
  font-size: 1.5rem;
  padding: 0.7rem 3rem;
  border-radius: var(--border-radius);
  cursor: pointer;
  background-color: #fff;
  border: 1px solid #e0e0e0;
`;
