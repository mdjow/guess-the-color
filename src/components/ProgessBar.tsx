import styled from "styled-components";

interface ProgressBarProps {
  progress: number;
}

export const ProgressBar = styled.div<ProgressBarProps>`
  position: relative;
  height: 0.6rem;
  background-color: #f5f5f5;

  &::after {
    content: "";
    display: flex;
    width: ${(props) => props.progress}%;
    height: 100%;
    background-color: #4caf50;
    transition: width 1s linear;
    background-color: 0.2s ease;
  }
`;
