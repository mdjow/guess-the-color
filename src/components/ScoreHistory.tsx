import styled from "styled-components";

type ScoreHistoryProps = {
  timeRemaining: number;
  score: number;
  highScore: number;
  onReset: () => void;
};

export const ScoreHistory = (props: ScoreHistoryProps) => (
  <ScoreHistoryWrapper>
    <BoxRow>
      <span>Remaning Time (s):</span>
      <span>{props.timeRemaining}</span>
    </BoxRow>
    <RestartButton onClick={props.onReset}>Restart</RestartButton>
    <BoxColumn>
      <BoxRow>
        <span>High score:</span>
        <span>{props.highScore}</span>
      </BoxRow>
      <BoxRow>
        <span>Score:</span>
        <span>{props.score}</span>
      </BoxRow>
    </BoxColumn>
  </ScoreHistoryWrapper>
);

const ScoreHistoryWrapper = styled.div`
  display: flex;
  width: 100%;
  background-color: #f5f5f5;
  border: 1px solid #e0e0e0;
  border-radius: 0.2rem;

  > *:not(:last-child) {
    border-right: 1px solid #e0e0e0;
  }
`;

const BoxColumn = styled.div`
  display: flex;
  flex-direction: column;
  font-weight: 300;
  width: 100%;
`;

const BoxRow = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  box-sizing: border-box;

  width: 100%;
  padding: 0.2rem 0.5rem;
  font-weight: 400;

  :nth-child(2) {
    font-weight: 700;
    font-size: 1.4rem;
  }

  &:nth-child(2) {
    border-top: 1px solid #e0e0e0;
  }
`;

const RestartButton = styled.button`
  width: 100%;
  padding: 0.3rem;
  font-size: 1.2rem;
  border: none;
  cursor: pointer;
  background-color: #e0e0e0;

  &:disabled {
    opacity: 0.5;
    cursor: not-allowed;
  }
`;
