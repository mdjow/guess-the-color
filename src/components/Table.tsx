import styled from "styled-components";

interface CellProps {
  colspan?: number;
}

export const Table = styled.div`
  display: flex;
  flex-direction: column;
`;

export const Row = styled.div`
  display: flex;
  justify-content: space-between;
  padding: 1rem 0;
  border-bottom: 1px solid var(--border-color);
`;

export const Cell = styled.div<CellProps>`
  flex: ${(props) => props.colspan ?? 1};
  text-align: center;
  padding: 0 0.5rem;
`;
