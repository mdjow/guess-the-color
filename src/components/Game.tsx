import { useContext, useEffect, useRef } from "react";

import styled from "styled-components";

import { ColorOptions } from "./ColorOptions";
import { ColorPreview } from "./ColorPreview";
import { ProgressBar } from "./ProgessBar";
import { ScoreHistory } from "./ScoreHistory";
import { GameContext } from "../contexts/GameContext";

const TIME_ANSWER_REMAINING = 10000;

export const Game = () => {
  const {
    state,
    startGame,
    resetData,
    resetGame,
    pickColor,
    decrementTime,
    endGame,
  } = useContext(GameContext);

  const {
    started,
    score,
    highScore,
    currentColor,
    colorOptions,
    timeGameRemaining,
  } = state;

  const timerGameRef = useRef<NodeJS.Timeout>();
  const timerAnswerRef = useRef<NodeJS.Timeout>();

  useEffect(() => {
    if (timeGameRemaining === 0) {
      clearInterval(timerGameRef.current);
      clearInterval(timerAnswerRef.current);
      endGame();
    }
  }, [endGame, timeGameRemaining]);

  const setTimerAnswerInterval = () => {
    if (timerAnswerRef.current) {
      clearInterval(timerAnswerRef.current);
    }

    timerAnswerRef.current = setInterval(
      () => pickColor(""),
      TIME_ANSWER_REMAINING
    );
  };

  const handleStartGame = () => {
    startGame();

    timerGameRef.current = setInterval(() => {
      decrementTime();
    }, 1000);

    setTimerAnswerInterval();
  };

  const handleResetGame = () => {
    clearInterval(timerGameRef.current);
    clearInterval(timerAnswerRef.current);

    resetGame();
  };

  const handleResetData = () => {
    clearInterval(timerGameRef.current);
    clearInterval(timerAnswerRef.current);

    resetData();
  };

  const handlePickColor = (color: string) => {
    setTimerAnswerInterval();

    pickColor(color);
  };

  return (
    <GameWrapper>
      <Title>Guess the color</Title>
      <ScoreHistory
        timeRemaining={timeGameRemaining}
        score={score}
        highScore={highScore}
        disabled={!started}
        onReset={handleResetGame}
      />
      <ProgressBar progress={timeGameRemaining / 300} />
      <ColorPreview
        color={currentColor}
        onStart={handleStartGame}
        started={started}
      />
      <ColorOptions
        options={colorOptions}
        disabled={!started}
        onSelect={handlePickColor}
      />
      <ResetDataButton data-testid="reset-all-data" onClick={handleResetData}>
        Reset all data
      </ResetDataButton>
    </GameWrapper>
  );
};

const GameWrapper = styled.main`
  display: flex;
  flex-direction: column;
  margin: auto;
  gap: 1.3rem;
  width: 40rem;
  padding: 3rem;
`;

const Title = styled.h1`
  font-size: 2.5rem;
  font-weight: 800;
  text-align: center;
`;

const ResetDataButton = styled.a`
  position: absolute;
  bottom: 1rem;
  right: 1rem;
  cursor: pointer;
  text-decoration: underline;
`;
