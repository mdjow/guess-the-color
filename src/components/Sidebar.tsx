import { useContext } from "react";

import styled from "styled-components";
import { CheckCircleOutlined, HighlightOff } from "@mui/icons-material";

import { TableCell, TableRow, Table, TableBody, TableHeader } from "./Table";
import { calculateTextColor } from "../utils/calculateTextColor";
import { GameAnswer, GameContext } from "../context/gameContext";
import { ScoreEnum } from "../enum/score";

export const Sidebar = () => {
  const { state } = useContext(GameContext);

  const buildColorCells = (answer: GameAnswer) => {
    if (!answer.pick) {
      return (
        <TableCell colSpan={2}>
          <TimeOutBox>Time Out</TimeOutBox>
        </TableCell>
      );
    }
    if (answer.pick === answer.color) {
      return (
        <TableCell colSpan={2}>
          <ColorBox color={answer.pick} />
        </TableCell>
      );
    }

    return (
      <>
        <TableCell>
          <ColorBox color={answer.pick} />
        </TableCell>
        <TableCell>
          <ColorBox color={answer.color} />
        </TableCell>
      </>
    );
  };

  return (
    <SidebarWrapper>
      <Title> Current/Latest Game</Title>
      <Table>
        <TableHeader>
          <TableCell>Guessed color</TableCell>
          <TableCell>Correct color</TableCell>
          <TableCell>Score</TableCell>
        </TableHeader>
        <TableBody>
          {state?.answers.map((answer) => (
            <TableRow key={`${answer.pick}-${answer.color}`}>
              {buildColorCells(answer)}
              <TableCell>
                <Score>
                  {answer.score === ScoreEnum.SUCCESS ? (
                    <CheckCircleOutlined color="success" />
                  ) : (
                    <HighlightOff color="error" />
                  )}
                  {answer.score}
                </Score>
              </TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </SidebarWrapper>
  );
};

const SidebarWrapper = styled.div`
  display: flex;
  flex-direction: column;
  width: 20rem;
  border-right: 1px solid var(--border-color);
`;

const Title = styled.h1`
  font-size: 1.5rem;
  font-weight: 600;
  text-align: center;
`;

interface ColorBoxProps {
  color?: string;
}

const ColorBox = styled.div<ColorBoxProps>`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  padding: 0.2rem;
  border-radius: var(--border-radius);
  background-color: ${(props) => props.color};

  &::after {
    content: "${(props) => props.color}";
    color: ${(props) => props.color && calculateTextColor(props.color)};
  }
`;

const TimeOutBox = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  padding: 0.2rem;
  border-radius: var(--border-radius);
  border: 1px solid #e0e0e0;
`;

const Score = styled.div`
  display: flex;
  justify-content: space-evenly;
  align-items: center;
  font-weight: 600;
`;
