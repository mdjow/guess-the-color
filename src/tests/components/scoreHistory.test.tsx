import { fireEvent, render } from "@testing-library/react";

import { ScoreHistory } from "../../components/ScoreHistory";

describe("ScoreHistory component", () => {
  const mockOnReset = jest.fn();

  const mockProps = {
    onReset: mockOnReset,
    timeRemaining: 30000,
    highScore: 100,
    score: 50,
    disabled: false,
  };

  test("should render component matches snapshot", () => {
    const { container } = render(<ScoreHistory {...mockProps} />);

    expect(container).toMatchSnapshot();
  });

  test("should not trigger event onReset if disabled", () => {
    const { getByText } = render(
      <ScoreHistory {...mockProps} disabled={true} />
    );

    fireEvent.click(getByText("Restart"));

    expect(mockOnReset).not.toHaveBeenCalled();
  });

  test("should trigger event onReset when called", () => {
    const { getByText } = render(<ScoreHistory {...mockProps} />);

    fireEvent.click(getByText("Restart"));

    expect(mockOnReset).toHaveBeenCalled();
  });
});
