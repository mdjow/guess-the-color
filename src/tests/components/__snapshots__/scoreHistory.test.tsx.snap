// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScoreHistory component should render component matches snapshot 1`] = `
<div>
  <div
    class="sc-beyTiQ jqMfbd"
  >
    <div
      class="sc-dmyDmy kbhgAm"
      direction="column"
    >
      <span>
        Remaning Time (s):
      </span>
      <span>
        30
      </span>
    </div>
    <button
      class="sc-hLQTFJ iQaexY"
      data-testid="reset-game"
    >
      Restart
    </button>
    <div
      class="sc-guDMob cwozNf"
    >
      <div
        class="sc-dmyDmy iZUnsQ"
      >
        <span>
          High score:
        </span>
        <span>
          100
        </span>
      </div>
      <div
        class="sc-dmyDmy iZUnsQ"
      >
        <span>
          Score:
        </span>
        <span>
          50
        </span>
      </div>
    </div>
  </div>
</div>
`;
